frame/support/src/traits.rs

```diff
@@ -653,9 +653,14 @@ bitmask! {
 	}
 }
 
+/// A trait to query the time of the current block.
 pub trait Time {
+	/// Type used for expressing timestamp.
 	type Moment: SimpleArithmetic + Parameter + Default + Copy;
 
+	/// Return the last timestamp set by inherent. If called before the inherent providing
```

Comment:
Hmm, why do we have the trait when it exactly replicates the behavior of the `timestamp` pallet? I would assume that time it something that strictly increases between blocks and could be in a simple case just the block number.

The documentation should go above the implementation of the `Time` trait for the `timestamp` pallet.

Here we should have a generic description what this trait provides you. IMHO.

Reply:
it is used by staking module to compute the duration of an era.

So I wanted to make the documentation of this trait explicit enough so that is it usable by staking module as is, without enforcing a specific implementation of this trait.

Currently staking module do call this trait during on_initialize because it is called by session module during on_initialize. (thus the duration of the era is slightly off actually)

Also I just realized staking module expect to have this value expressed as millisecond, which is not enforced anywhere in those traits. maybe we should bound more on this trait so that Moment::year() returns the value for a duration of one year. (or something else to get this information).

Reply:
This sounds completely implementation specific to the staking pallet. We should document that the `Time` type of staking expects milliseconds.

I'm not against documentation on this trait, but when we put all these constraints, we can just import `timestamp` and remove the trait. Than we are sure that all the constraints are fulfilled, otherwise anyone could provide *anything* that implements this trait.

Maybe we should just be more specific on what the `now` method returns. This means, we should maybe rename the `now` function to `now_milliseconds` or similar. Or provide multiple of these methods that could return an `Option`. 

I'm not even sure how the `staking` module can expect that something is milliseconds, while even the type for the time is generic. I think there is some huge miss-use with this trait.

We need to clearly improve, but just putting some documentation above the the time trait, will not solve the `staking` problem IMHO. Maybe we should have a more specific time trait that exposes functions like `now_milliseconds`, `now_seconds` etc, so that the implementer is aware what is expected.

Reply:
about the seconds/milliseconds issue yes, we can precise the unit of `now` or require a const in the trait which is `YEAR` or `HOUR` or `SECOND`.

for the issue about when now returns a correct value I don't maybe it is fine to put in the documentation of the implementation of timestamp that it is not available at genesis config and that before the timestamp inherent it return the time of previous block.
Then in staking module we can specify that Time is not called during genesis config. and for the computation of the era duration (which is slightly off) we can say accept that like an "imprecision" of the available clock (as we queryied just before it is updated), or we can also fix that and say in the doc of staking that Time is only called during on_finalize.